London | 25-SDC-July | Ali Qassab | Module Legacy Code | Sprint 1 | Can't log in from profile page #6

AliQassab opened the pull request:
- Redirect users to home page after logout instead of staying on profile page
- Resolves 501 error when logging in from profile page after logout
- Fixes #6

<!--

You must title your PR like this:

Region | Cohort | FirstName LastName | Sprint | Assignment Title

For example,

London | 25-ITP-May | Carol Owen | Sprint 1 | Alarm Clock

Fill in the template below - remove any sections that don't apply.

Complete the self checklist - replace each empty box in the checklist [ ] with a [x].

Add the label "Needs Review" and you will get review.

Respond to volunteer reviews until the volunteer marks it as "Complete".

-->

## Learners, PR Template

Self checklist

- [ ] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [ ] My changes meet the requirements of the task
- [ ] I have tested my changes
- [ ] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

**Problem:**
Users experienced a 501 "Server does not support this operation" error when:
Logging in → clicking a username → logging out → logging back in
Expected to see the profile page but got an error instead

**Root Cause:**
After logout, users remained on the profile page (#/profile/AS). When logging back in, the application tried to restore this profile page state, which triggered operations that caused the 501 error.

**Solution:**
Added navigation redirect to the logout function:

function logout() {
  state.destroyState();
  window.location.hash = "/";  // Redirect to home page
  return { success: true };
}

AliQassab committed:
Fix: Add logout navigation redirect to prevent 501 error

- Redirect users to home page after logout instead of staying on profile page
- Resolves 501 error when logging in from profile page after logout
- Fixes #6

illicitonion reviewed (commented):
This is a cute workaround, but the core of this bug report is that we're showing a component that doesn't work - if you go to http://localhost:8000/index.html#/profile/AS there's a login component that can never work (if you're not logged in).

Ideally we would make the login component work. If we can't do that, we should hide the component.

There are probably other pages that show a login form when you're not logged in, that don't work - there's probably a common underlying pattern here we need to fix.

AliQassab committed:
Fix login form on profile and hashtag pages

AliQassab commented:
Thanks for the feedback. I've fixed this.
1. Made the login component work on profile and hashtag pages
2. Hidden profile/hashtag content when not logged in
3. Found and fixed the same issue on hashtag pages

illicitonion commented:
The underlying bug looks fixed here, well done!

Mixing in the "hide content if you're not logged in", however, seems messy. Deciding to hide profile content if you're not logged in isn't related to the bug being fixed. Importantly, if we _do_ want to hide that profile content:
1. We should also add authentication to the backend, so people can't access it directly.
2. We probably want to show some UI about how the profile won't be shown while you're logged out.

I'll mark this as complete because the underlying bug has been fixed, but I'd recommend reverting the "if not logged in, hide" changes from this PR.

illicitonion commented:
Oh also, please update your PR description - the "solution" section is no longer accurate :)
